### PR TITLE
ARIA-254 Create multiple nodes per template

### PR DESCRIPTION
**Pull request by tliron**
* New aria.Scaling policy (and "scaling" role)
* NodeTemplate model no longer stores scaling values (default_instances,
etc.) but instead fetches them from applicable scaling policies
* Some code cleanup

### CODE REVIEW

File: tests/end2end/test_nodecellar.py [hunk 1]
```diff
@@ -39,4 +39,4 @@ def _verify_deployed_service_in_storage(service_name, model_storage):
     service = service_templates[0].services[service_name]
     assert service.name == service_name
     assert len(service.executions) == 0  # dry executions leave no traces
-    assert len(service.nodes) == 10
+    assert len(service.nodes) == 13
```
Comment (mxmrlv) [inline tests/end2end/test_nodecellar.py]: why 13? I though you added just one other instance

File: extensions/aria_extension_tosca/simple_v1_0/types.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from aria.utils.collections import FrozenDict, FrozenList
+from aria.utils.collections import (FrozenDict, FrozenList)
```
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/types.py]: if adding parentheses, why not try and follow the style guide? i.e. `from aria.utils import collections`

File: aria/modeling/service_template.py [hunk 3]
```diff
@@ -613,6 +597,56 @@ def dump(self):
...
+    @property
+    def default_instances(self):
+        # TODO: currently finds the first matching policy; but we should emit a validation error
+        # if more than one policy applies to the same node
+        for policy_template in self.service_template.policy_templates.itervalues():
+            if policy_template.type.role == 'scaling':
+                if policy_template.is_for_node_template(self.name):
+                    default_instances = policy_template.properties.get('default_instances')
+                    if (default_instances is not None) \
+                        and (default_instances.type_name == 'integer'):
```
Comment (mxmrlv) [inline aria/modeling/service_template.py]: how can it not be 'integer'?

File: tests/end2end/test_nodecellar.py (hunk 1 above)
Comment (tliron) [inline tests/end2end/test_nodecellar.py]: The policy is on the `node_cellar_group` which contains three 3 templates. So we get 3 extra nodes.

File: extensions/aria_extension_tosca/simple_v1_0/types.py (hunk 2 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/types.py]: I thought we agreed that that kind of style is a recommendation but not a requirement. If we switch to that style here it would make a lot of the parser code extremely verbose and frankly unreadable.

File: aria/modeling/service_template.py (hunk 3 above)
Comment (tliron) [inline aria/modeling/service_template.py]: We cannot expect anybody be to using the ARIA profile as is -- all we are relying here is that there is an attribute with that name. I am adding that extra validation of type to make sure that we don't cause breakage by somebody messing with the parameters. Note that we also do this validation in setting "default" attributes, like `tosca_name` and `tosca_id`.

File: extensions/aria_extension_tosca/simple_v1_0/types.py (hunk 2 above)
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/types.py]: if by adding the module prefix makes the code unreadable, maybe we need to refactor the structure?

File: tests/end2end/test_nodecellar.py (hunk 1 above)
Comment (mxmrlv) [inline tests/end2end/test_nodecellar.py]: kewl

File: extensions/aria_extension_tosca/profiles/tosca-simple-1.0/capabilities.yaml [hunk 4]
```diff
@@ -132,6 +132,7 @@ capability_types:
       specification: tosca-simple-1.0
       specification_section: 5.4.10
       specification_url: 'http://docs.oasis-open.org/tosca/TOSCA-Simple-Profile-YAML/v1.0/cos01/TOSCA-Simple-Profile-YAML-v1.0-cos01.html#DEFN_TYPE_CAPABILITIES_SCALABLE'
+      role: scaling
```
Comment (mxmrlv) [inline extensions/aria_extension_tosca/profiles/tosca-simple-1.0/capabilities.yaml]: what does it mean to have a scaling role?

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py [hunk 5]
```diff
@@ -162,6 +164,30 @@ def convert_capability_from_definition_to_assignment(context, presentation, cont
...
+    merge_raw_parameter_definitions(context, presentation, raw_properties,
+                                    capability_definition.properties, 'properties')
+
+    if raw_properties:
+        capability_definition._raw['properties'] = raw_properties
+        capability_definition._reset_method_cache()
+
+    # Merge occurrences
+    occurrences = from_capability_definition._raw.get('occurrences')
+    if (occurrences is not None) and (capability_definition._raw.get('occurrences') is None):
```
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: why the redundant parenthesis?

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py [hunk 6]
```diff
@@ -52,12 +52,13 @@ def get_inherited_capability_definitions(context, presentation, for_presentation
...
     parent = presentation._get_parent(context)
-    capability_definitions = get_inherited_capability_definitions(context, parent,
-                                                                  for_presentation=presentation) \
-                                                                  if parent is not None \
-                                                                  else OrderedDict()
+    capability_definitions = get_inherited_capability_definitions(
+        context, parent, for_presentation) if parent is not None else OrderedDict()
 
     # Add/merge our capability definitions
     our_capability_definitions = presentation.capabilities
```
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: it's really hard to follow with name like 
capability_definitions and our_capability_definitions

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py [hunk 7]
```diff
@@ -162,6 +164,30 @@ def convert_capability_from_definition_to_assignment(context, presentation, cont
     return CapabilityAssignment(name=presentation._name, raw=raw, container=container)
 
 
+def merge_capability_definition(context, presentation, capability_definition,
+                                from_capability_definition):
```
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: the hard to follow comment applies here as well

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py [hunk 8]
```diff
@@ -162,6 +164,30 @@ def convert_capability_from_definition_to_assignment(context, presentation, cont
...
+    raw_properties = OrderedDict()
+
+    # Merge properties from type
+    from_property_defintions = from_capability_definition.properties
+    merge_raw_parameter_definitions(context, presentation, raw_properties, from_property_defintions,
+                                    'properties')
+
+    # Merge our properties
+    merge_raw_parameter_definitions(context, presentation, raw_properties,
+                                    capability_definition.properties, 'properties')
```
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: if i get this right, capability_definitions is the "global" definition of this capability, while "from..." if the definitions for this instance. 
and we first apply our own definitions and only then try and merge the global one - while maintaining the idea that the more specific definitions will be the ones to hold?

File: aria/modeling/service_template.py [hunk 9]
```diff
@@ -690,19 +666,104 @@ def dump(self):
...
-                else ' or more'))
             utils.dump_dict_values(self.properties, 'Properties')
             utils.dump_dict_values(self.attributes, 'Attributes')
             utils.dump_interfaces(self.interface_templates)
             utils.dump_dict_values(self.artifact_templates, 'Artifact templates')
             utils.dump_dict_values(self.capability_templates, 'Capability templates')
             utils.dump_list_values(self.requirement_templates, 'Requirement templates')
 
+    @property
+    def next_index(self):
```
Comment (mxmrlv) [inline aria/modeling/service_template.py]: cannot remain public

File: aria/modeling/service_template.py [hunk 10]
```diff
@@ -690,19 +666,104 @@ def dump(self):
...
+        :rtype: int
+        """
+
+        max_index = 0
+        if self.nodes:
+            max_index = max(int(n.name.rsplit('_', 1)[-1]) for n in self.nodes)
+        return max_index + 1
+
+    @property
+    def next_name(self):
```
Comment (mxmrlv) [inline aria/modeling/service_template.py]: so is this

Comment (mxmrlv) [inline aria/modeling/service_template.py]: All the new functions are seem to be for orchestrator use only. Why would the user need to use these?

File: aria/modeling/service_template.py [hunk 11]
```diff
@@ -690,19 +666,104 @@ def dump(self):
...
+                    extract_properties(policy_template.properties)
+
+        # Defaults
+        default_property('min_instances', 0)
+        default_property('max_instances', 1)
+        default_property('default_instances', 1)
+
+        # Validate
+        # pylint: disable=too-many-boolean-expressions
+        if (scaling['min_instances'] < 0) or \
```
Comment (mxmrlv) [inline aria/modeling/service_template.py]: may any? and remove the pylint issue?

File: aria/modeling/service_template.py [hunk 12]
```diff
@@ -690,19 +666,104 @@ def dump(self):
...
+            if name not in scaling:
+                scaling[name] = value
+
+        # From our scaling capabilities
+        for capability_template in self.capability_templates.itervalues():
+            if capability_template.type.role == 'scaling':
+                extract_properties(capability_template.properties)
+
+        # From service scaling policies
+        for policy_template in self.service_template.policy_templates.itervalues():
```
Comment (mxmrlv) [inline aria/modeling/service_template.py]: so are we overriding the more specific properties with the policy based properties?

File: extensions/aria_extension_tosca/profiles/tosca-simple-1.0/capabilities.yaml (hunk 4 above)
Comment (tliron) [inline extensions/aria_extension_tosca/profiles/tosca-simple-1.0/capabilities.yaml]: The `role` meta data is the way to specify special behavior for types in ARIA. This is much safer than checking if name == `aria.Scaling`, or even if it inherits from that. Roles are inherited with the type, so subtypes will have the same role (unless they override it). We use this in a few cases, for example there is `host` role for Compute and for HostedOn relationships.

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 6 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: I tend to agree. :( This is some very tricky code that has gone through a lot of changes. I do want to fix this (there are a bunch of related files in this module) but I don't think this PR is the right place.

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 5 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: I do not consider them redundant at all. Nobody should have to remember operation precedence, you should be able to tell at a glance. I recommend everybody follow this practice in every language. :)

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 7 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: +1, but I suggest refactoring some other time

File: aria/modeling/service_template.py (hunk 9 above)
Comment (tliron) [inline aria/modeling/service_template.py]: Actually, this was intentional -- I thought it could be useful for outside callers. But no problem changed this and also `_next_name` similarly.

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 8 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: Not exactly. This is a utility method called by `get_inherited_capability_definitions`. Its goal it to merge the current node type's capability definitions over the inherited parent node type's capability definitions. "from" here is the parent's.

File: aria/modeling/service_template.py (hunk 12 above)
Comment (tliron) [inline aria/modeling/service_template.py]: No, the opposite. There is no overriding: the `extract_property` function will not extract if the value is already there. First to come wins. :)

File: aria/modeling/service_template.py (hunk 11 above)
Comment (tliron) [inline aria/modeling/service_template.py]: I'm not sure how `any` would work here. How would you write this?

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 5 above)
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: well i do agree when it comes to an elaborate conditions :)

File: extensions/aria_extension_tosca/profiles/tosca-simple-1.0/capabilities.yaml (hunk 4 above)
Comment (mxmrlv) [inline extensions/aria_extension_tosca/profiles/tosca-simple-1.0/capabilities.yaml]: :+1:

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 8 above)
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: I do get the logic of the entire method, i was referring to the order of the `merge_raw_parameter_definitions`...

File: aria/modeling/service_template.py (hunk 12 above)
Comment (mxmrlv) [inline aria/modeling/service_template.py]: Oh, didn't notice it. I would think that there is not real need to check, just override the inherited properties. But this is my personal preference

File: aria/modeling/service_template.py (hunk 11 above)
Comment (mxmrlv) [inline aria/modeling/service_template.py]:
```
if any([
 scaling['min_instances'] < 0,
 scaling['max_instances'] < 0,
 scaling['default_instances'] < 0,
 scaling['max_instances'] < scaling['min_instances'],
 scaling['default_instances'] < scaling['min_instances'],
 scaling['default_instances'] > scaling['max_instances']
]):
...
```

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 7 above)
Comment (mxmrlv) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: well, i might go with it regarding the first naming issue, but this is a new function, why not rename its arguments properly?

File: aria/modeling/service_template.py (hunk 11 above)
Comment (tliron) [inline aria/modeling/service_template.py]: I see no difference in readability between this and using `or`, seems to me just to be bowing to PyLint's weirdness... but I don't mind switching.

File: extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py (hunk 7 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/capabilities.py]: Because I prefer it to fit with the style of the rest of code right now, to keep it consistent. I would like to change it, but everything together and give it some more thought.
